flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java

```diff
@@ -126,56 +144,71 @@ public ZooKeeperStateHandleStore(
+            throws PossibleInconsistentStateException, Exception {
         checkNotNull(pathInZooKeeper, "Path in ZooKeeper");
         checkNotNull(state, "State");
 
         final String path = normalizePath(pathInZooKeeper);
 
         RetrievableStateHandle<T> storeHandle = storage.store(state);
 
-        boolean success = false;
+        byte[] serializedStoreHandle = serializeStateHandle(storeHandle);
 
+        // initialize flag to serve the failure case
+        boolean discardState = true;
         try {
-            // Serialize the state handle. This writes the state to the backend.
-            byte[] serializedStoreHandle = InstantiationUtil.serializeObject(storeHandle);
-
-            // Write state handle (not the actual state) to ZooKeeper. This is expected to be
-            // smaller than the state itself. This level of indirection makes sure that data in
-            // ZooKeeper is small, because ZooKeeper is designed for data in the KB range, but
-            // the state can be larger.
-            // Create the lock node in a transaction with the actual state node. That way we can
-            // prevent
-            // race conditions with a concurrent delete operation.
-            client.inTransaction()
-                    .create()
-                    .withMode(CreateMode.PERSISTENT)
-                    .forPath(path, serializedStoreHandle)
-                    .and()
-                    .create()
-                    .withMode(CreateMode.EPHEMERAL)
-                    .forPath(getLockPath(path))
-                    .and()
-                    .commit();
-
-            success = true;
+            writeStoreHandleTransactionally(path, serializedStoreHandle);
+            discardState = false;
             return storeHandle;
-        } catch (KeeperException.NodeExistsException e) {
+        } catch (Exception e) {
+            if (indicatesPossiblyInconsistentState(e)) {
+                discardState = false;
+                throw new PossibleInconsistentStateException(e);
+            }
+
             // We wrap the exception here so that it could be caught in DefaultJobGraphStore
-            throw new AlreadyExistException("ZooKeeper node " + path + " already exists.", e);
+            throw ExceptionUtils.findThrowable(e, KeeperException.NodeExistsException.class)
+                    .map(
+                            nee ->
+                                    new AlreadyExistException(
+                                            "ZooKeeper node " + path + " already exists.", nee))
+                    .orElseThrow(() -> e);
         } finally {
-            if (!success) {
-                // Cleanup the state handle if it was not written to ZooKeeper.
-                if (storeHandle != null) {
-                    storeHandle.discardState();
-                }
+            if (discardState) {
+                storeHandle.discardState();
```

Comment:
I think we could get rid of a boolean flag and simplify code a bit by discarding the handle in the `catch` block instead of `finally`, something like this:
```
try {
  writeStoreHandleTransactionally();
} catch (Exception e) {
  if (indicatesPossiblyInconsistentState) {
    throw PossibleInconsistentStateException;
  } else {
    storeHandle.discardState();
  }
}
```
WDYT?

Reply:
That's a good catch. Thanks for the idea. I will apply it. 👍